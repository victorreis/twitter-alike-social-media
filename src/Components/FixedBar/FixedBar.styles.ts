import styled from 'styled-components';

import { FixedBarStyleProps } from './FixedBar.types';

export const FixedBarContainer = styled.div<FixedBarStyleProps>`
  display: flex;
  height: 3rem;
  top: 0;
  left: 0;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  align-items: center;

  box-sizing: border-box;
  border-bottom-width: 1px;
  border-bottom-style: solid;
  border-bottom-color: ${({ theme }) => theme.colors.background.default.normal};

  background-color: ${({ theme }) => theme.colors.background.default.darkest};
  color: ${({ theme }) => theme.colors.font.default};
  ${({ theme }) => theme.typographies.h3};

  justify-content: ${({ justifyContent }) => justifyContent};
`;
