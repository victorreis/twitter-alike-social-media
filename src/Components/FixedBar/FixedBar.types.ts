import { TestProps } from '../../Config/Tests/Test.types';

export const availableJustifyContents = [
  'flex-start',
  'flex-end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
] as const;
export type JustifyContent = typeof availableJustifyContents[number];

export interface RequiredFixedBarProps {
  /**
   * Component's children.
   */
  children: React.ReactNode;
}

export interface DefaultFixedBarProps {
  /**
   * Sets the content justification.
   * @default 'flex-start'.
   */
  justifyContent?: JustifyContent;
}

export interface OptionalFixedBarProps {}

export type FixedBarProps = RequiredFixedBarProps &
  DefaultFixedBarProps &
  OptionalFixedBarProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;

export type FixedBarStyleProps = Required<DefaultFixedBarProps>;
