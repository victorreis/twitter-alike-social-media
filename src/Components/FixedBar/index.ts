export * from './FixedBar.types';
export * from './FixedBar';
