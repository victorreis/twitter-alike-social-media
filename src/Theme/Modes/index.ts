export { darkTheme } from './Dark.theme';
