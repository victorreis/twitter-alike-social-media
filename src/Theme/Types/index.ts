export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
