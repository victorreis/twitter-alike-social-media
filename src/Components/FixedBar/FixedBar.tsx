import { TestProps } from '../../Config/Tests/Test.types';
import { FixedBarContainer } from './FixedBar.styles';
import { FixedBarProps, DefaultFixedBarProps } from './FixedBar.types';

export const fixedBarDefaults: Required<DefaultFixedBarProps> &
  Required<TestProps> = {
  testID: 'FixedBar',
  justifyContent: 'flex-start',
};

export const FixedBar: React.FC<FixedBarProps> = (props): JSX.Element => {
  const {
    testID = fixedBarDefaults.testID,
    children,
    justifyContent = fixedBarDefaults.justifyContent,
    style,
    ...others
  } = props;

  return (
    <FixedBarContainer
      data-testid={testID}
      justifyContent={justifyContent}
      style={style}
      {...others}
    >
      {children}
    </FixedBarContainer>
  );
};
