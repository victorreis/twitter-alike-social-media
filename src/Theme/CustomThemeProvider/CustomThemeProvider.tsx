/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
