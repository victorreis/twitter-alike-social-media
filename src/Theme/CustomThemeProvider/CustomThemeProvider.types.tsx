import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
