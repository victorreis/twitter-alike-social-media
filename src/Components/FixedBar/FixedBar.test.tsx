import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb } from '../../Utils/Transform/hexToRgb.util';
import { FixedBar, fixedBarDefaults } from './FixedBar';
import {
  RequiredFixedBarProps,
  FixedBarProps,
  JustifyContent,
} from './FixedBar.types';

describe('fixedBar component tests', () => {
  const text = 'text';
  const newJustifyContent: JustifyContent = 'space-between';

  const requiredProps: RequiredFixedBarProps = {
    children: text,
  };

  const setup = (props?: FixedBarProps) => {
    const renderRTR = () =>
      renderRTRCreator<FixedBarProps>(FixedBar, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<FixedBarProps>(FixedBar, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(fixedBarDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text when passed as children`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(text);

      expect(element).toBeInTheDocument();
    });

    it(`should render '${fixedBarDefaults.justifyContent}' as the default 'justifyContent'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        justifyContent: fixedBarDefaults.justifyContent,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'justifyContent' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        justifyContent: newJustifyContent,
      }).renderRTR().root;
      const element = instance.findByProps({
        justifyContent: newJustifyContent,
      });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(fixedBarDefaults.testID);

      expect(container).toHaveStyle({
        borderBottomStyle: 'solid',
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.darkest
        ),
        color: hexToRgb(themes.default.colors.font.default),
        justifyContent: fixedBarDefaults.justifyContent,
        ...themes.default.typographies.h3,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
