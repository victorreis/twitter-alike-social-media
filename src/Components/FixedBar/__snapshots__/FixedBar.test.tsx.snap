// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixedBar component tests snapshot tests should render correctly 1`] = `
<div
  className="FixedBarContainer-twitter-alike-social-media__sc-1hwkltx-0 geprJN"
  data-testid="FixedBar"
>
  text
</div>
`;
